<a href="https://colab.research.google.com/github/Pragna235/Python_Projects/blob/main/Pandas_DataFrame_14.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Instructions

#### Goal of the Project

This project is designed for you to practice and solve the activities that are based on the concepts covered in the following lessons:

1. While Loop, Data-Type Conversion and Conditional Statements.

2. Pandas DataFrame.


---

#### Getting Started:

1. Click on this link to open the Colab file for this project.

  https://colab.research.google.com/drive/1GO3d4EFI1qq70Ae-fDZ89243k3lVn-tn?usp=sharing 

2. Create a duplicate copy of the Colab file as described below.

  - Click on the **File menu**. A new drop-down list will appear.

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/lesson-0/0_file_menu.png' width=500>

  - Click on the **Save a copy in Drive** option. A duplicate copy will get created. It will open up in the new tab on your web browser.

  <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/lesson-0/1_create_colab_duplicate_copy.png' width=500>

3. After creating the duplicate copy of the notebook, please rename it in the **YYYY-MM-DD_StudentName_Project14** format.

4. Now, write your code in the prescribed code cells.


---

### Activities


  

#### Activity 1: Create Pandas DataFrame

Create a pandas DataFrame by using the below link which has the dataset of NBA players in the CSV format: 

**Dataset** : https://media.geeksforgeeks.org/wp-content/uploads/nba.csv

Also, print the first five rows of the dataset.

In [ ]:
# Create a pandas DataFrame
import pandas as pd
nba_players_df=pd.read_csv("https://media.geeksforgeeks.org/wp-content/uploads/nba.csv")
print(nba_players_df)
nba_players_df.head()

              Name            Team  ...            College     Salary
0    Avery Bradley  Boston Celtics  ...              Texas  7730337.0
1      Jae Crowder  Boston Celtics  ...          Marquette  6796117.0
2     John Holland  Boston Celtics  ...  Boston University        NaN
3      R.J. Hunter  Boston Celtics  ...      Georgia State  1148640.0
4    Jonas Jerebko  Boston Celtics  ...                NaN  5000000.0
..             ...             ...  ...                ...        ...
453   Shelvin Mack       Utah Jazz  ...             Butler  2433333.0
454      Raul Neto       Utah Jazz  ...                NaN   900000.0
455   Tibor Pleiss       Utah Jazz  ...                NaN  2900000.0
456    Jeff Withey       Utah Jazz  ...             Kansas   947276.0
457            NaN             NaN  ...                NaN        NaN

[458 rows x 9 columns]


Name            Team  Number  ... Weight            College     Salary
0  Avery Bradley  Boston Celtics     0.0  ...  180.0              Texas  7730337.0
1    Jae Crowder  Boston Celtics    99.0  ...  235.0          Marquette  6796117.0
2   John Holland  Boston Celtics    30.0  ...  205.0  Boston University        NaN
3    R.J. Hunter  Boston Celtics    28.0  ...  185.0      Georgia State  1148640.0
4  Jonas Jerebko  Boston Celtics     8.0  ...  231.0                NaN  5000000.0

[5 rows x 9 columns]

---

#### Activity 1.2 : Check for Null values

Check for missing values in the DataFrame.

In [ ]:
# Check for missing values in the above Pandas series.
nba_players_df.isnull()

Name   Team  Number  Position    Age  Height  Weight  College  Salary
0    False  False   False     False  False   False   False    False   False
1    False  False   False     False  False   False   False    False   False
2    False  False   False     False  False   False   False    False    True
3    False  False   False     False  False   False   False    False   False
4    False  False   False     False  False   False   False     True   False
..     ...    ...     ...       ...    ...     ...     ...      ...     ...
453  False  False   False     False  False   False   False    False   False
454  False  False   False     False  False   False   False     True   False
455  False  False   False     False  False   False   False     True   False
456  False  False   False     False  False   False   False    False   False
457   True   True    True      True   True    True    True     True    True

[458 rows x 9 columns]

**Q:** Are there any missing values in the DataFrame ?

**A:** Yes, there some missing values in the dataframe.

---

#### Activity 1.3 : Get the Number of rows and columns

Get the number of rows and columns of the DataFrame.

In [ ]:
# Get the Number of rows and columns
nba_players_df.shape

(458, 9)

**Q:** How many rows are there in the DataFrame?

**A:** There are 458 rows in the dataframe.

---

#### Activity 1.4: Create Pandas series for the player 'Avery Bradley'

Using the `iloc[]` function, create a Pandas series for the player Avery Bradley and store it in a variable called `avery_bradley`.

In [ ]:
# Create pandas series for the player 'Avery Bradley'
avery_bradley=nba_players_df.iloc[0,:]
avery_bradley

Name         Avery Bradley
Team        Boston Celtics
Number                   0
Position                PG
Age                     25
Height                 6-2
Weight                 180
College              Texas
Salary         7.73034e+06
Name: 0, dtype: object

**Q**: Which team does Avery Bradely plays for?

**A**: Boston Celtics.

---

#### Activity 2: Find the number of solutions for a Quadratic equation

A quadratic equation $ax² + b x + c = 0$ has either 0, 1, or 2 distinct solutions for real values of $x$. Given $a$, $b$ and $c$, you should only return the number of solutions that we can get from the equation.

The above quadratic equation is solved using the formula, \begin{array}{*{20}c} {x = \frac{{ - b \pm \sqrt {b^2 - 4ac} }}{{2a}}} & {{\rm{when}}} & {ax^2 + bx + c = 0} \\ \end{array}

The `±` in the above formula means there are **TWO** answers.

Do you see $b^2 − 4ac$ in the formula above? It is called the **Discriminant**, because it can "discriminate" between the possible types of answer:

  - when $b^2 − 4ac$ is **positive**, we get **two** Real solutions.

  - when $b^2 − 4ac$ is **zero**, we get just **one** real solution. 

  - when $b^2 − 4ac$ is **negative**, we get a pair of **complex solution**.


Examples:
```
1. solutions(1, 0, -1) ➞ 2
# x² - 1 = 0 has two solutions (x = 1 and x = -1).

2. solutions(1, 0, 1) ➞ 0
# x² + 1 = 0 has no solutions.
```

**Note**: You do not need to calculate the solutions, just return how many solutions are there by calculating the value of $b^2 − 4ac$.


In [ ]:
# Number of solutions for a quadratic equation
def solutions(a,b,c):
  
  if (b**2-(4*a*c))>0:
    print("2 solutions")
  elif (b**2-(4*a*c))==0:
    print("1 solution")
  else:
    print("No solutions")
solutions(1,0,-1)

2 solutions


---

### Submitting the Project:

1. After finishing the project, click on the **Share** button on the top right corner of the notebook. A new dialog box will appear.

  <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/2_share_button.png' width=500>

2. In the dialog box, make sure that '**Anyone on the Internet with this link can view**' option is selected and then click on the **Copy link** button.

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/3_copy_link.png' width=500>

3. The link of the duplicate copy (named as **YYYY-MM-DD_StudentName_Project14**) of the notebook will get copied 

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/4_copy_link_confirmation.png' width=500>

4. Go to your dashboard and click on the **My Projects** option.
   
   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/5_student_dashboard.png' width=800>

  <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/6_my_projects.png' width=800>

5. Click on the **View Project** button for the project you want to submit.

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/7_view_project.png' width=800>

6. Click on the **Submit Project Here** button.

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/8_submit_project.png' width=800>

7. Paste the link to the project file named as **YYYY-MM-DD_StudentName_Project14** in the URL box and then click on the **Submit** button.

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/9_enter_project_url.png' width=800> 